bson-depth-tracking/DESIGN.md

```diff
@@ -0,0 +1,24 @@
+# BSON Depth Tracking
+To help ensure that users of the BSON library do not crash their applications do a stack overflow
+from attempting to decode a malicious document, we need to add depth tracking to the BSON library’s
+validation functions and allow users to set the maximum depth of a BSON document the library will
+decode.
+
+## Scope of Work
+The `bson.Document`, `bson.Reader`, `bson.Element`, and `bson.Value` types need to have their
+validate methods augmented with additional depth tracking parameters to track how deep the
+validation should go. Each of these types has a `validate` method except for the `bson.Value` type.
+
+All of the internal calls during the validation process should use the `validate` method. This
+method will take the current depth and the maximum depth. An increase in the current depth number
+will occur before validating a BSON element of type `0x03`, `0x04`, or `0x0F`.  If, when entering
```

Comment:
Most people don't know what these are. 0x03 is a document, etc... might as well just spell it out.